keps/sig-node/3063-dynamic-resource-allocation/README.md

```diff
@@ -0,0 +1,2346 @@
+depending on a flag in the ResourceClaim spec.
+
+Immediate allocation is useful when allocating a resource is expensive (for
+example, programming an FPGA) and the resource therefore is meant to be used by
+multiple different Pods, either in parallel or one after the other. The
+downside is that Pod resource requirements cannot be considered when choosing
+where to allocate. If a resource was allocated so that it is only available on
+one node and the Pod cannot run there because other resources like RAM or CPU
+are exhausted on that node, then the Pod cannot run elsewhere. The same applies
+to resources that are available on a certain subset of the nodes and those
+nodes are busy.
+
+Delayed allocation solves this by integrating allocation with Pod scheduling:
+an attempt to schedule a Pod triggers allocation of pending resources for nodes
+that the scheduler has deemed suitable. Scheduling the pod is then put on hold
+until all resources are allocated. This avoids scenarios where a Pod is
+permanently assigned to a node which can't fit the pod because of the pod's
+other resource requirements.
+
+### Sharing a single ResourceClaim
+
+Pods reference resource claims in a new PodSpec.ResourceClaims list. Each
+resource in that list can then be made available to one or more containers in
+that Pod. Depending on the capabilities defined in the
+ResourceClaim by the driver, a ResourceClaim can be used exclusively
+by one pod at a time or an unlimited number of pods. Support for additional
+constraints (maximum number of pods, maximum number of nodes) could be
+added once there are use cases for those.
+
+Users of a ResourceClaim don't need to be Pods. This KEP specifically supports
+Pods as users and describes how kube-scheduler and kubelet will deal with Pods
+that depend on a ResourceClaim, but the API and some custom resource driver
+might also be useful for controllers to manage resources without using those
+resources for Pods.
+
+### Ephemeral vs. persistent ResourceClaims lifecycle
+
+A persistent ResourceClaim has a lifecyle that is independent of any particular
+pod. It gets created and deleted by the user. This is useful for resources
+which are expensive to configure and that can be used multiple times by pods,
+either at the same time or one after the other. Such persistent ResourceClaims
+get referenced in the pod spec by name. When a PodTemplateSpec in an app
+controller spec references a ResourceClaim by name, all pods created by that
+controller also use that name and thus share the resources allocated for that
+ResourceClaim.
+
+But often, each Pod is meant to have exclusive access to its own ResourceClaim
+instance instead. To support such ephemeral resources without having to modify
+all controllers that create Pods, an entry in the new PodSpec.ResourceClaims
+list can also be a ResourceClaimTemplate. When a Pod gets created, such a
+template will be used to create a normal ResourceClaim with the Pod as owner
```

Comment:
I swear I commented on this but I just can't find it.

I don't quite follow this.

1) User creates a deployment with a claimTemplate for an ExpensiveResource
2) Deployment creates a pod with a claimTemplate
3) Someone (who?) converts that claimTemplate into a Claim (with a fixed name)
4) blah blah blah scheduler, ExpensiveResource is allocated
5) kubelet runs the pod
6) pod eventually finishes
7) kubelet marks pod "Succeeded"

At this point, the pod exists and is not being deleted.  It has a claim which points to a valid Claim object.  The ExpensiveResource is still "owned" by the claim (not released).

This state might exist for hours or days.  ownerRef isn't doing anything useful.  Is someone going to delete the "ephemeral" claim?

Reply:
> At this point, the pod exists and is not being deleted. 

It isn't? I was under the impression that Deployment's goal was to ensure a certain number of running pods. If one container stops with a zero exit code, will the Deployment really not do something about this?

This was not something that came up when reviewing generic ephemeral inline volumes, which use the same approach. So either it was missed during review and hasn't bitten a user yet, or the pod does get deleted. I'll have to try that.

> This state might exist for hours or days. ownerRef isn't doing anything useful. Is someone going to delete the "ephemeral" claim?

Not in the current design. But it also makes sense: if the pod still exists, perhaps there is a way how the container might get restarted? Then the resource would be needed for it again.

Reply:
>> At this point, the pod exists and is not being deleted.
>
> It isn't? I was under the impression that Deployment's goal was to ensure a certain number of running pods. If one container stops with a zero exit code, will the Deployment really not do something about this?

I made a bad example - Deployments don't "finish".  But Jobs do.  Will update my text. 

>> This state might exist for hours or days. ownerRef isn't doing anything useful. Is someone going to delete the "ephemeral" claim?
>
> Not in the current design. But it also makes sense: if the pod still exists, perhaps there is a way how the container might get restarted? Then the resource would be needed for it again.

I bet these all have volumes that are racking up cloud bills while the pods sit there.  And you know, for a volume I can almost see the reason (debug, etc).  For a GPU or something, I am not sure.